File: code/125-Valid Palindrome/125-Valid Palindrome.py
"""
LeetCode 125 - Valid Palindrome

解题步骤（自动整理）
1. 将起点/初始层入队，并用 visited 或原地标记避免重复访问
2. 按层或按队列弹出元素扩展相邻节点，满足条件时更新答案
3. 队列处理完毕后返回结果（或在首次到达目标时提前返回）
"""
# Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.
#
# Note: For the purpose of this problem, we define empty string as valid palindrome.
#
# Example 1:
# Input: "A man, a plan, a canal: Panama"
# Output: true
#
# Example 2:
# Input: "race a car"
# Output: false

class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        deque = []
        for char in s:
            if char.isalpha() or char.isdigit():
                deque.append(char.lower())
        while len(deque) > 1:
            if deque.pop(0) != deque.pop(-1):
                return False
        return True

File: code/125-Valid Palindrome/125-Valid Palindrome.c
/*
LeetCode 125 - Valid Palindrome

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
*/
// A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

// Given a string s, return true if it is a palindrome, or false otherwise.

//

// Example 1:

// Input: s = "A man, a plan, a canal: Panama"
// Output: true
// Explanation: "amanaplanacanalpanama" is a palindrome.
// Example 2:

// Input: s = "race a car"
// Output: false
// Explanation: "raceacar" is not a palindrome.
// Example 3:

// Input: s = " "
// Output: true
// Explanation: s is an empty string "" after removing non-alphanumeric characters.
// Since an empty string reads the same forward and backward, it is a palindrome.
//

// Constraints:

// 1 <= s.length <= 2 * 10^5
// s consists only of printable ASCII characters.

bool validLetter(char c) {
    int low = c - 'a';
    int high = c - 'A';
    int digit = c - '0';
    return (low >=0 && low <= 25) || (high >= 0 && high <= 25) || (digit >= 0 && digit <= 9);
}

bool sameLetter(char a, char b) {
    int lowa = a - 'a', lowb = b - 'a', higha = a - 'A', highb = b - 'A';
    if (a == b){
        return true;
    } else if (lowa >= 0 && lowa <= 25 && lowa == highb) {
        return true;
    } else if (higha >= 0 && higha <= 25 && higha == lowb) {
        return true;
    } else {
        return false;
    }

}

bool isPalindrome(char * s){
    int left = 0, right = strlen(s) - 1;
    while (left < right) {
        while (!validLetter(s[left]) && left < right) {
            left ++;
        }
        while (!validLetter(s[right]) && left < right) {
            right --;
        }
        if (!sameLetter(s[left], s[right])) {
            return false;
        } else {
            left ++;
            right --;
        }
    }
    return true;
}
